### PR TITLE
Upgrade to Mapbox GL JS

### PR DESCRIPTION
**Pull request by brangi**
## Description

https://github.com/department-of-veterans-affairs/va.gov-team/issues/14786

## Testing done
updated test cases - https://github.com/department-of-veterans-affairs/vets-website/pull/14749

## Screenshots
TBD

## Tasks
- [x] Update to GL JS library
- [x] Create method to restrict from Prod
- [x] Test in review instance

## Acceptance Criteria

Following existing functionality will be maintained
- [x] Search within existing parameters is successful, including return of no results
- [x] Pins for locations
- [x] "I am here" indicator
- [x] Uses existing 80 mile bounding box

---

- [x] Performance within acceptable parameters
- [x] Mobile experience maintained
- [x] Changes are restricted from Prod
- [x] FL does not autosearch if not results are found within 80 miles 

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/facility-locator/routes.jsx [hunk 1]
```diff
@@ -2,18 +2,37 @@ import FacilityLocatorApp from './containers/FacilityLocatorApp';
...
+  const isStaging = environment.isStaging();
+  const isCypress = window.Cypress;
+
+  if (isUnitTest || isCypress) return VAMap;
+
+  if (isReviewEnvironment || isLocal || isStaging) {
+    return FacilitiesMap;
+  }
+  return VAMap;
+};
```
Comment (mpelzsherman) [inline src/applications/facility-locator/routes.jsx:28]: 👍  - though at some point we will need to figure out how to write Cypress tests against the new map.

Comment (brangi) [reply src/applications/facility-locator/routes.jsx:29]: https://github.com/department-of-veterans-affairs/vets-website/pull/14749, however, it should be merged after the map upgrade, it does not make any sense supporting test cases against 2 different map UI

Comment (mpelzsherman) [reply src/applications/facility-locator/routes.jsx:29]: Have you tried running the current Cypress tests against the new Map?

File: src/applications/facility-locator/containers/FacilitesMap.jsx [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+      window.removeEventListener('resize', debouncedResize);
+    };
+  }, []);
+
+  const updateUrlParams = params => {
+    const { location, currentQuery } = props;
+
+    const queryParams = {
+      ...location.query,
+      zoomLevel: currentQuery.zoomLevel,
```
Comment (mpelzsherman) [inline src/applications/facility-locator/containers/FacilitesMap.jsx:105]: Do we still need a `zoomLevel` in the query? I thought we were using fitBounds to determine the zoom level.

File: src/applications/facility-locator/containers/FacilitesMap.jsx [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+            page: resultsPage,
+          });
+          setIsSearching(false);
+        }
+        if (searchResultTitleRef.current) {
+          setFocus(searchResultTitleRef.current);
+        }
+      }
+      return () => {
+        // remove the listeners
```
Comment (mpelzsherman) [inline src/applications/facility-locator/containers/FacilitesMap.jsx:147]: Let's not forget to do this!

File: src/applications/facility-locator/containers/FacilitesMap.jsx [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    });
+  };
+
+  const setupMap = (setMapInit, mapContainerInit) => {
+    mapboxgl.accessToken = mapboxToken;
+    const mapInit = new mapboxgl.Map({
+      container: mapContainerInit,
+      style: 'mapbox://styles/mapbox/outdoors-v11',
+      center: [-99.27246093750001, 40.17887331434698], // Initial state search query reducer
+      zoom: 3,
```
Comment (mpelzsherman) [inline src/applications/facility-locator/containers/FacilitesMap.jsx:238]: If we need to specify a hard-coded Zoom level, can we try making it very high, so fitBounds adjusts it as needed?

File: src/applications/facility-locator/containers/FacilitesMap.jsx [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+        letter: markersLetters.next().value,
+      };
+      const markerElement = buildMarker('location', { loc, attrs });
+      new mapboxgl.Marker(markerElement)
+        .setLngLat([loc.attributes.long, loc.attributes.lat])
+        .addTo(map);
+    });
+
+    if (props.currentQuery.bounds) {
+      map.fitBounds(props.currentQuery.bounds); // {duration: 0} to disable animation
```
Comment (mpelzsherman) [inline src/applications/facility-locator/containers/FacilitesMap.jsx:185]: I think we want to use the **markers** to determine the bounds, not the query.

Here is some (untested) code to try:

```js
const bounds = new mapboxgl.LngLatBounds();

locations.forEach(location => {
    bounds.extend(new mapboxgl.LngLat(location.latitude, location.longitude);
});

map.fitBounds(bounds);
```

Comment (mpelzsherman) [reply src/applications/facility-locator/containers/FacilitesMap.jsx:185]: Actually you could probably compute the marker bounds in the same loop where we are adding the markers to the map.

File: src/applications/facility-locator/containers/FacilitesMap.jsx [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+      page,
+    });
+  };
+
+  const setupMap = (setMapInit, mapContainerInit) => {
+    mapboxgl.accessToken = mapboxToken;
+    const mapInit = new mapboxgl.Map({
+      container: mapContainerInit,
+      style: 'mapbox://styles/mapbox/outdoors-v11',
+      center: [-99.27246093750001, 40.17887331434698], // Initial state search query reducer
```
Comment (mpelzsherman) [inline src/applications/facility-locator/containers/FacilitesMap.jsx:237]: Can we move those coordinates into a constant?

File: src/applications/facility-locator/routes.jsx (hunk 1 above)
Comment (brangi) [reply src/applications/facility-locator/routes.jsx:29]: yes but needs fixes, that's why I have #14749.

File: src/applications/facility-locator/containers/FacilitesMap.jsx (hunk 4 above)
Comment (brangi) [reply src/applications/facility-locator/containers/FacilitesMap.jsx:253]: these values is o that it centers the map on the us at initial render, I will put them in constants.

Comment (mpelzsherman) [reply src/applications/facility-locator/containers/FacilitesMap.jsx:253]: Thanks. As I mentioned elsewhere, those same coordinates are used in several places, so yes, please put them in a const.

**Review by rianfowler — APPROVED** [review]

File: src/applications/facility-locator/containers/FacilitesMap.jsx [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+            page: resultsPage,
+          });
+          setIsSearching(false);
+        }
+        if (searchResultTitleRef.current) {
+          setFocus(searchResultTitleRef.current);
+        }
+      }
+      return () => {
+        // remove the listeners
```
Comment (mpelzsherman) [inline src/applications/facility-locator/containers/FacilitesMap.jsx:146]: remove comment?

File: src/applications/facility-locator/containers/FacilitesMap.jsx [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+
+    const queryStringObj = appendQuery(
+      `/find-locations${location.pathname}`,
+      queryParams,
+    );
+
+    browserHistory.push(queryStringObj);
+  };
+
+  useEffect(
```
Comment (mpelzsherman) [inline src/applications/facility-locator/containers/FacilitesMap.jsx:120]: I think it's standard practice to group all of the `useEffect` instances together.

File: src/applications/facility-locator/containers/FacilitesMap.jsx (hunk 7 above)
Comment (brangi) [reply src/applications/facility-locator/containers/FacilitesMap.jsx]: done

File: src/applications/facility-locator/containers/FacilitesMap.jsx (hunk 8 above)
Comment (brangi) [reply src/applications/facility-locator/containers/FacilitesMap.jsx]: done

File: src/applications/facility-locator/sass/facility-locator.scss [hunk 9]
```diff
@@ -485,13 +485,16 @@ $facility-locator-color-gray: #ccc;
...
-  border-radius: 50%;
+  justify-content: center;
+  align-content: center;
+  display: inline-flex;
   font-size: 16px;
-  font-family: 'Source Sans Pro', sans-serif !important;
-}
+  width: 23px;
+  height: 23px;;
+ }
```
Comment (mpelzsherman) [inline src/applications/facility-locator/sass/facility-locator.scss:497]: I think the following changes will help the letters appear more centered:

```css
font-size: 15px;
padding-top: 1px;
width: 24px;
height: 24px;
```

File: src/applications/facility-locator/utils/helpers.jsx [hunk 10]
```diff
@@ -14,6 +15,47 @@ export const setFocus = selector => {
   }
 };
 
+export const clearLocationMarkers = () => {
+  const locationMarkers = window.document.getElementsByClassName(
+    'mapboxgl-marker',
+  );
+  for (let i = locationMarkers.length - 1; i >= 0; i--) {
+    locationMarkers[i].remove();
+  }
```
Comment (mpelzsherman) [inline src/applications/facility-locator/utils/helpers.jsx:24]:
```suggestion
  Array.from(locationMarkers).forEach(marker => marker.remove());
```

Comment (mpelzsherman) [reply src/applications/facility-locator/utils/helpers.jsx:24]: Note that `locationMarkers` is an `HTMLCollection` and it does not have the `forEach` method. But I still think this is slightly preferable to the C-style for loop.

**Review by mpelzsherman — APPROVED** [review]
Suggestions are optional. Looks great! Let's get this out on staging and see what people think.

**Pull request merged**